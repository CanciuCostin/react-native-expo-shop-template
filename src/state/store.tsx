import { configureStore } from "@reduxjs/toolkit";
import productsReducer from "./productsDataSlice";

export const store = configureStore({
  reducer: {
    productsData: productsReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
