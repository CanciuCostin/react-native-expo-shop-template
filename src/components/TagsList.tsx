import React from 'react';
import {
  FlatList,
  SafeAreaView,
  StyleSheet,
  TouchableOpacity,
  Text,
} from 'react-native';
import { ProductTag } from '@models/types';
import { AppDispatch, RootState } from '@state/store';
import { useDispatch, useSelector } from 'react-redux';
import { setSelectedTag } from '@state/productsDataSlice';

const styles = StyleSheet.create({
  container: {
    height: 30,
  },
  item: {
    paddingHorizontal: 10,
    backgroundColor: 'red',
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 10,
  },
  title: {},
});

export default function TagsList() {
  const tags = useSelector((state: RootState) => state.productsData.tags);
  const dispatch = useDispatch<AppDispatch>();

  const renderTagItem = ({ item }: { item: ProductTag }) => (
    <TouchableOpacity
      onPress={() => dispatch(setSelectedTag(item.tagId))}
      style={styles.item}
    >
      <Text style={styles.title}>{item.tagName}</Text>
    </TouchableOpacity>
  );

  return (
    <SafeAreaView style={styles.container}>
      <FlatList
        horizontal={true}
        data={tags}
        renderItem={renderTagItem}
        showsHorizontalScrollIndicator={false}
        keyExtractor={(item) => `${item.tagId}`}
      />
    </SafeAreaView>
  );
}
