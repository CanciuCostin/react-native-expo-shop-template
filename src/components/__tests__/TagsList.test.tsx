import React from 'react';
import { render } from '@testing-library/react-native';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import TagsList from '@components/TagsList';
import { RootState } from '@state/store';
import {
  defaultTagsInitialState,
  emptyInitialState,
  longTagNameInitialState,
  specialCharactersTagsInitialState,
} from '@test-utils/mockStore';

const mockStore = configureStore([]);

describe('TagsList Component', () => {
  it('renders correctly with given tags', () => {
    const initialState: RootState = {
      productsData: {
        tags: [
          { tagId: 1, tagName: 'React' },
          { tagId: 2, tagName: 'JavaScript' },
          { tagId: 3, tagName: 'Expo' },
        ],
        products: [],
        categories: [],
        selectedCategoryId: 0,
        selectedProductId: 0,
        selectedTags: [],
      },
    };
    const store = mockStore(initialState);

    const { getByText } = render(
      <Provider store={store}>
        <TagsList />
      </Provider>,
    );

    expect(getByText('React')).toBeTruthy();
    expect(getByText('JavaScript')).toBeTruthy();
    expect(getByText('Expo')).toBeTruthy();
  });

  it('renders empty list correctly', () => {
    const store = mockStore(emptyInitialState);

    const { queryByText } = render(
      <Provider store={store}>
        <TagsList />
      </Provider>,
    );

    expect(queryByText('React')).toBeNull();
    expect(queryByText('JavaScript')).toBeNull();
    expect(queryByText('Expo')).toBeNull();
  });

  it('renders tags with special characters correctly', () => {
    const store = mockStore(specialCharactersTagsInitialState);

    const { getByText } = render(
      <Provider store={store}>
        <TagsList />
      </Provider>,
    );

    expect(getByText('Tag1+')).toBeTruthy();
    expect(getByText('Tag2#')).toBeTruthy();
    expect(getByText('Tag3/')).toBeTruthy();
  });

  it('renders tags with long names correctly', () => {
    const store = mockStore(longTagNameInitialState);

    const { getByText } = render(
      <Provider store={store}>
        <TagsList />
      </Provider>,
    );

    expect(
      getByText('A very long tag name to test rendering in the component'),
    ).toBeTruthy();
  });

  it('matches snapshot', () => {
    const store = mockStore(defaultTagsInitialState);

    const component = render(
      <Provider store={store}>
        <TagsList />
      </Provider>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders the correct number of tags', () => {
    const store = mockStore(defaultTagsInitialState);

    const { getAllByText } = render(
      <Provider store={store}>
        <TagsList />
      </Provider>,
    );

    const tags = getAllByText(/.*/);
    expect(tags).toHaveLength(3);
  });
});
