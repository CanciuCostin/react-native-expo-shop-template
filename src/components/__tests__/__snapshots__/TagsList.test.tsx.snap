// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsList Component matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignSelf": "center",
      "height": 66.5,
    }
  }
>
  <RCTScrollView
    data={
      [
        {
          "tagId": "1",
          "tagName": "Tag 1",
        },
        {
          "tagId": "2",
          "tagName": "Tag 2",
        },
        {
          "tagId": "3",
          "tagName": "Tag 3",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgb(255, 255, 255)",
              "borderColor": "rgb(216, 216, 216)",
              "borderRadius": 10,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginHorizontal": 7.5,
              "opacity": 1,
              "paddingHorizontal": 22.5,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "rgb(28, 28, 30)",
                  "fontSize": 24,
                  "fontWeight": "normal",
                },
                {},
              ]
            }
            testID="custom-text"
          >
            Tag 1
          </Text>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgb(255, 255, 255)",
              "borderColor": "rgb(216, 216, 216)",
              "borderRadius": 10,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginHorizontal": 7.5,
              "opacity": 1,
              "paddingHorizontal": 22.5,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "rgb(28, 28, 30)",
                  "fontSize": 24,
                  "fontWeight": "normal",
                },
                {},
              ]
            }
            testID="custom-text"
          >
            Tag 2
          </Text>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={
          [
            {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgb(255, 255, 255)",
              "borderColor": "rgb(216, 216, 216)",
              "borderRadius": 10,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginHorizontal": 7.5,
              "opacity": 1,
              "paddingHorizontal": 22.5,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "rgb(28, 28, 30)",
                  "fontSize": 24,
                  "fontWeight": "normal",
                },
                {},
              ]
            }
            testID="custom-text"
          >
            Tag 3
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
